From bc93f55a60fffd49bf6129ebac5bb2a0b0e92e05 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Wed, 16 Jun 2021 18:13:46 -0700
Subject: [PATCH 119/196] [tests] Ensure block height, amounts, and metadata
 hash are correct (#195)

When modifying various aspects of the consensus rules, many tests were
updated explicitly to set these fields correctly. However, this wasn't
done for all tests and missed due to them being part of the extended
suite. They were discovered when timing were updated in ABC 0.23.3. This
commit added height to the `create_block` signature, uses
`prepare_block` and updates test to ensure that the appropriate output
from coinbases are used.
---
 test/functional/abc-invalid-chains.py         |   3 +-
 .../abc-mempool-coherence-on-activations.py   |   3 +-
 test/functional/abc-minimaldata.py            |   3 +-
 test/functional/abc-replay-protection.py      |   3 +-
 test/functional/abc-schnorr.py                |   3 +-
 test/functional/abc-schnorrmultisig.py        |   3 +-
 test/functional/abc-sync-chain.py             |   3 +-
 test/functional/abc-transaction-ordering.py   |   5 +-
 test/functional/abc_mining_basic.py           |   1 -
 test/functional/abc_p2p_compactblocks.py      |  20 +-
 test/functional/abc_p2p_fullblocktest.py      |  37 +--
 test/functional/example_test.py               |   5 +-
 test/functional/feature_assumevalid.py        |  12 +-
 test/functional/feature_bip68_sequence.py     |   3 +-
 test/functional/feature_block.py              | 244 ++++++++----------
 test/functional/feature_cltv.py               |  12 +-
 test/functional/feature_csv_activation.py     |   3 +-
 test/functional/feature_dersig.py             |   3 +-
 test/functional/feature_pruning.py            |   2 +-
 .../logos_enforce_standard_consensus.py       |   3 +-
 test/functional/logos_feature_int63.py        |   3 +-
 test/functional/logos_feature_minerfund.py    |  12 +-
 test/functional/mempool_updatefromblock.py    |   7 +-
 test/functional/p2p_compactblocks.py          |   3 +-
 test/functional/p2p_fingerprint.py            |   3 +-
 test/functional/p2p_invalid_block.py          |  12 +-
 test/functional/p2p_invalid_tx.py             |   3 +-
 test/functional/p2p_sendheaders.py            |  25 +-
 test/functional/p2p_tx_download.py            |   6 +-
 test/functional/p2p_unrequested_blocks.py     |  30 +--
 test/functional/rpc_blockchain.py             |   3 +-
 test/functional/test_framework/blocktools.py  |   4 +-
 test/functional/test_framework/messages.py    |   3 +
 .../wallet_resendwallettransactions.py        |   3 +-
 34 files changed, 193 insertions(+), 295 deletions(-)

diff --git a/test/functional/abc-invalid-chains.py b/test/functional/abc-invalid-chains.py
index 1f68be5fb..a526b39b9 100755
--- a/test/functional/abc-invalid-chains.py
+++ b/test/functional/abc-invalid-chains.py
@@ -40,8 +40,7 @@ class InvalidChainsTest(BitcoinTestFramework):
         coinbase = create_coinbase(height, CScript([self.counter]))
         coinbase.rehash()
         self.counter += 1
-        block = create_block(base_block_hash, coinbase, block_time)
-        block.nHeight = height
+        block = create_block(base_block_hash, coinbase, height, block_time)
         prepare_block(block)
         self.tip = block
         self.block_heights[block.sha256] = height
diff --git a/test/functional/abc-mempool-coherence-on-activations.py b/test/functional/abc-mempool-coherence-on-activations.py
index bdb999f04..6ed7f53d1 100755
--- a/test/functional/abc-mempool-coherence-on-activations.py
+++ b/test/functional/abc-mempool-coherence-on-activations.py
@@ -146,8 +146,7 @@ class MempoolCoherenceOnActivationsTest(BitcoinTestFramework):
         height = self.block_heights[base_block_hash] + 1
         coinbase = create_coinbase(height)
         coinbase.rehash()
-        block = create_block(base_block_hash, coinbase, block_time)
-        block.nHeight = height
+        block = create_block(base_block_hash, coinbase, height, block_time)
         prepare_block(block)
 
         # Do PoW, which is cheap on regnet
diff --git a/test/functional/abc-minimaldata.py b/test/functional/abc-minimaldata.py
index d9f8b4f99..52acd5a0d 100755
--- a/test/functional/abc-minimaldata.py
+++ b/test/functional/abc-minimaldata.py
@@ -86,8 +86,7 @@ class MinimaldataTest(BitcoinTestFramework):
         block_time = (parent.nTime + 1) if nTime is None else nTime
 
         block = create_block(
-            parent.sha256, create_coinbase(block_height), block_time)
-        block.nHeight = block_height
+            parent.sha256, create_coinbase(block_height), block_height, block_time)
         block.vtx.extend(transactions)
         prepare_block(block)
         block.solve()
diff --git a/test/functional/abc-replay-protection.py b/test/functional/abc-replay-protection.py
index 74fbb171a..d8d67d06e 100755
--- a/test/functional/abc-replay-protection.py
+++ b/test/functional/abc-replay-protection.py
@@ -77,8 +77,7 @@ class ReplayProtectionTest(BitcoinTestFramework):
         height = self.block_heights[base_block_hash] + 1
         coinbase = create_coinbase(height)
         coinbase.rehash()
-        block = create_block(base_block_hash, coinbase, block_time)
-        block.nHeight = height
+        block = create_block(base_block_hash, coinbase, height, block_time)
         prepare_block(block)
         self.tip = block
         self.block_heights[block.sha256] = height
diff --git a/test/functional/abc-schnorr.py b/test/functional/abc-schnorr.py
index 1da5456fb..5f7130d83 100755
--- a/test/functional/abc-schnorr.py
+++ b/test/functional/abc-schnorr.py
@@ -101,8 +101,7 @@ class SchnorrTest(BitcoinTestFramework):
         block_time = (parent.nTime + 1) if nTime is None else nTime
 
         block = create_block(
-            parent.sha256, create_coinbase(block_height), block_time)
-        block.nHeight = block_height
+            parent.sha256, create_coinbase(block_height), block_height, block_time)
         block.vtx.extend(transactions)
         prepare_block(block)
         self.block_heights[block.sha256] = block_height
diff --git a/test/functional/abc-schnorrmultisig.py b/test/functional/abc-schnorrmultisig.py
index 5008e4337..d8ad08fde 100755
--- a/test/functional/abc-schnorrmultisig.py
+++ b/test/functional/abc-schnorrmultisig.py
@@ -104,8 +104,7 @@ class SchnorrMultisigTest(BitcoinTestFramework):
         block_time = (parent.nTime + 1) if nTime is None else nTime
 
         block = create_block(
-            parent.sha256, create_coinbase(block_height), block_time)
-        block.nHeight = block_height
+            parent.sha256, create_coinbase(block_height), block_height, block_time)
         block.vtx.extend(transactions)
         prepare_block(block)
         block.solve()
diff --git a/test/functional/abc-sync-chain.py b/test/functional/abc-sync-chain.py
index 4a850c5fa..9b25a3663 100755
--- a/test/functional/abc-sync-chain.py
+++ b/test/functional/abc-sync-chain.py
@@ -49,8 +49,7 @@ class SyncChainTest(BitcoinTestFramework):
 
         blocks = []
         for i in range(NUM_IBD_BLOCKS * 2):
-            block = create_block(tip, create_coinbase(height), time)
-            block.nHeight = height
+            block = create_block(tip, create_coinbase(height), height, time)
             prepare_block(block)
             blocks.append(block)
             tip = block.sha256
diff --git a/test/functional/abc-transaction-ordering.py b/test/functional/abc-transaction-ordering.py
index 997436005..f301d3faf 100755
--- a/test/functional/abc-transaction-ordering.py
+++ b/test/functional/abc-transaction-ordering.py
@@ -67,12 +67,12 @@ class TransactionOrderingTest(BitcoinTestFramework):
         coinbase.rehash()
         if spend is None:
             # We need to have something to spend to fill the block.
-            block = create_block(base_block_hash, coinbase, block_time)
+            block = create_block(base_block_hash, coinbase, height, block_time)
         else:
             # Burn half of the fees
             coinbase.vout[0].nValue += spend.tx.vout[spend.n].nValue // 2
             coinbase.rehash()
-            block = create_block(base_block_hash, coinbase, block_time)
+            block = create_block(base_block_hash, coinbase, height, block_time)
 
             # Make sure we have plenty enough to spend going forward.
             spendable_outputs = deque([spend])
@@ -117,7 +117,6 @@ class TransactionOrderingTest(BitcoinTestFramework):
         if tx_count > 0:
             assert_equal(len(block.vtx), tx_count)
 
-        block.nHeight = height
         block.update_size()
         block.rehash_extended_metadata()
         block.solve()
diff --git a/test/functional/abc_mining_basic.py b/test/functional/abc_mining_basic.py
index ab6cd2cdd..764d33898 100755
--- a/test/functional/abc_mining_basic.py
+++ b/test/functional/abc_mining_basic.py
@@ -73,7 +73,6 @@ class AbcMiningRPCTest(BitcoinTestFramework):
         assert_equal(node.getmempoolinfo()['size'], 0)
         share_amount = SUBSIDY * COIN // 26
         block_template_data = node.getblocktemplate()
-        print(block_template_data)
         expected_funding_outputs = 13
         miner_fund_outputs = block_template_data['coinbasetxn']['minerfund']['outputs']
         assert_equal(len(miner_fund_outputs), 13)
diff --git a/test/functional/abc_p2p_compactblocks.py b/test/functional/abc_p2p_compactblocks.py
index 860e33536..a8a6c11fb 100755
--- a/test/functional/abc_p2p_compactblocks.py
+++ b/test/functional/abc_p2p_compactblocks.py
@@ -20,6 +20,7 @@ from test_framework.blocktools import (
     create_coinbase,
     create_tx_with_script,
     make_conform_to_ctor,
+    prepare_block,
 )
 from test_framework.cdefs import ONE_MEGABYTE
 from test_framework.messages import (
@@ -135,12 +136,12 @@ class FullBlockTest(BitcoinTestFramework):
         if spend is None:
             # We need to have something to spend to fill the block.
             assert_equal(block_size, 0)
-            block = create_block(base_block_hash, coinbase, block_time)
+            block = create_block(base_block_hash, coinbase, height, block_time)
         else:
             # all but one satoshi to fees
-            coinbase.vout[0].nValue += spend.tx.vout[spend.n].nValue - 1
+            coinbase.vout[0].nValue += (spend.tx.vout[spend.n].nValue - 1) // 2
             coinbase.rehash()
-            block = create_block(base_block_hash, coinbase, block_time)
+            block = create_block(base_block_hash, coinbase, height, block_time)
 
             # Make sure we have plenty enough to spend going forward.
             spendable_outputs = deque([spend])
@@ -150,7 +151,7 @@ class FullBlockTest(BitcoinTestFramework):
                 tx = CTransaction()
                 # Spend from one of the spendable outputs
                 spend = spendable_outputs.popleft()
-                tx.vin.append(CTxIn(COutPoint(spend.tx.sha256, spend.n)))
+                tx.vin.append(CTxIn(COutPoint(spend.tx.txid, spend.n)))
                 # Add spendable outputs
                 for i in range(4):
                     tx.vout.append(CTxOut(0, CScript([OP_TRUE])))
@@ -216,7 +217,6 @@ class FullBlockTest(BitcoinTestFramework):
 
             # Now that we added a bunch of transaction, we need to recompute
             # the merkle root.
-            make_conform_to_ctor(block)
             block.hashMerkleRoot = block.calc_merkle_root()
 
         # Check that the block size is what's expected
@@ -224,7 +224,7 @@ class FullBlockTest(BitcoinTestFramework):
             assert_equal(len(block.serialize()), block_size)
 
         # Do PoW, which is cheap on regnet
-        block.solve()
+        prepare_block(block)
         self.tip = block
         self.block_heights[block.sha256] = height
         assert number not in self.blocks
@@ -246,11 +246,7 @@ class FullBlockTest(BitcoinTestFramework):
 
         # get an output that we previously marked as spendable
         def get_spendable_output():
-            return PreviousSpendableOutput(spendable_outputs.pop(0).vtx[0], 0)
-
-        # move the tip back to a previous block
-        def tip(number):
-            self.tip = self.blocks[number]
+            return PreviousSpendableOutput(spendable_outputs.pop(0).vtx[0], 1)
 
         # shorthand for functions
         block = self.next_block
@@ -348,7 +344,7 @@ class FullBlockTest(BitcoinTestFramework):
         tx.rehash()
         b2.vtx[0] = tx
         b2.hashMerkleRoot = b2.calc_merkle_root()
-        b2.solve()
+        prepare_block(b2)
 
         # Now we create the compact block and send it
         comp_block = HeaderAndShortIDs()
diff --git a/test/functional/abc_p2p_fullblocktest.py b/test/functional/abc_p2p_fullblocktest.py
index 0a0c4936a..751f6a22e 100755
--- a/test/functional/abc_p2p_fullblocktest.py
+++ b/test/functional/abc_p2p_fullblocktest.py
@@ -20,6 +20,7 @@ from test_framework.blocktools import (
     create_coinbase,
     create_tx_with_script,
     make_conform_to_ctor,
+    prepare_block,
 )
 from test_framework.cdefs import (
     ONE_MEGABYTE,
@@ -37,6 +38,7 @@ from test_framework.script import (
     CScript,
     OP_RETURN,
     OP_TRUE,
+    OP_PUSHDATA1,
 )
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import assert_equal
@@ -60,7 +62,8 @@ class FullBlockTest(BitcoinTestFramework):
         self.blocks = {}
         self.excessive_block_size = 100 * ONE_MEGABYTE
         self.extra_args = [['-whitelist=noban@127.0.0.1',
-                            "-excessiveblocksize={}".format(self.excessive_block_size)]]
+                            "-excessiveblocksize={}".format(self.excessive_block_size), 
+                            '-acceptnonstdtxn=1']]
         self.supports_cli = False
         # The default timeout is not enough when submitting large blocks with
         # TSAN enabled
@@ -91,16 +94,15 @@ class FullBlockTest(BitcoinTestFramework):
         # First create the coinbase
         height = self.block_heights[base_block_hash] + 1
         coinbase = create_coinbase(height)
-        coinbase.rehash()
         if spend is None:
             # We need to have something to spend to fill the block.
             assert_equal(block_size, 0)
-            block = create_block(base_block_hash, coinbase, block_time)
+            block = create_block(base_block_hash, coinbase, height, block_time)
         else:
             # all but one satoshi to fees
-            coinbase.vout[0].nValue += spend.tx.vout[spend.n].nValue - 1
+            coinbase.vout[0].nValue += (spend.tx.vout[spend.n].nValue - 1) // 2
             coinbase.rehash()
-            block = create_block(base_block_hash, coinbase, block_time)
+            block = create_block(base_block_hash, coinbase, height, block_time)
 
             # Make sure we have plenty engough to spend going forward.
             spendable_outputs = deque([spend])
@@ -110,7 +112,7 @@ class FullBlockTest(BitcoinTestFramework):
                 tx = CTransaction()
                 # Spend from one of the spendable outputs
                 spend = spendable_outputs.popleft()
-                tx.vin.append(CTxIn(COutPoint(spend.tx.sha256, spend.n)))
+                tx.vin.append(CTxIn(COutPoint(spend.tx.txid, spend.n)))
                 # Add spendable outputs
                 for i in range(4):
                     tx.vout.append(CTxOut(0, CScript([OP_TRUE])))
@@ -171,7 +173,6 @@ class FullBlockTest(BitcoinTestFramework):
 
             # Now that we added a bunch of transaction, we need to recompute
             # the merkle root.
-            make_conform_to_ctor(block)
             block.hashMerkleRoot = block.calc_merkle_root()
 
         # Check that the block size is what's expected
@@ -179,6 +180,7 @@ class FullBlockTest(BitcoinTestFramework):
             assert_equal(len(block.serialize()), block_size)
 
         # Do PoW, which is cheap on regnet
+        prepare_block(block)
         block.solve()
         self.tip = block
         self.block_heights[block.sha256] = height
@@ -200,29 +202,12 @@ class FullBlockTest(BitcoinTestFramework):
 
         # get an output that we previously marked as spendable
         def get_spendable_output():
-            return PreviousSpendableOutput(spendable_outputs.pop(0).vtx[0], 0)
+            return PreviousSpendableOutput(spendable_outputs.pop(0).vtx[0], 1)
 
         # move the tip back to a previous block
         def tip(number):
             self.tip = self.blocks[number]
 
-        # adds transactions to the block and updates state
-        def update_block(block_number, new_transactions):
-            block = self.blocks[block_number]
-            self.add_transactions_to_block(block, new_transactions)
-            old_sha256 = block.sha256
-            make_conform_to_ctor(block)
-            block.hashMerkleRoot = block.calc_merkle_root()
-            block.solve()
-            # Update the internal state just like in next_block
-            self.tip = block
-            if block.sha256 != old_sha256:
-                self.block_heights[
-                    block.sha256] = self.block_heights[old_sha256]
-                del self.block_heights[old_sha256]
-            self.blocks[block_number] = block
-            return block
-
         # shorthand for functions
         block = self.next_block
 
@@ -257,7 +242,7 @@ class FullBlockTest(BitcoinTestFramework):
         # Reject oversized blocks with bad-blk-length error
         block(18, spend=out[17], block_size=self.excessive_block_size + 1)
         node.p2p.send_blocks_and_test(
-            [self.tip], node, success=False, reject_reason='bad-blk-length')
+            [self.tip], node, success=False, force_send=True, reject_reason='bad-blk-size')
 
         # Rewind bad block.
         tip(17)
diff --git a/test/functional/example_test.py b/test/functional/example_test.py
index e5066d552..698fb8c38 100755
--- a/test/functional/example_test.py
+++ b/test/functional/example_test.py
@@ -179,13 +179,13 @@ class ExampleTest(BitcoinTestFramework):
         height = self.nodes[0].getblockcount()
 
         for i in range(10):
+            height += 1
             # Use the mininode and blocktools functionality to manually build a block
             # Calling the generate() rpc is easier, but this allows us to exactly
             # control the blocks and transactions.
             block = create_block(
                 self.tip, create_coinbase(
-                    height + 1), self.block_time)
-            block.nHeight = height + 1
+                    height), height, self.block_time)
             prepare_block(block)
             block_message = msg_block(block)
             # Send message is used to send a P2P message to the node over our
@@ -194,7 +194,6 @@ class ExampleTest(BitcoinTestFramework):
             self.tip = block.sha256
             blocks.append(self.tip)
             self.block_time += 1
-            height += 1
 
         self.log.info(
             "Wait for node1 to reach current tip (height 11) using RPC")
diff --git a/test/functional/feature_assumevalid.py b/test/functional/feature_assumevalid.py
index 46a113bab..fea61b964 100755
--- a/test/functional/feature_assumevalid.py
+++ b/test/functional/feature_assumevalid.py
@@ -122,8 +122,7 @@ class AssumeValidTest(BitcoinTestFramework):
         # Create the first block with a coinbase output to our key
         height = 1
         block = create_block(self.tip, create_coinbase(
-            height, coinbase_pubkey), self.block_time)
-        block.nHeight = height
+            height, coinbase_pubkey), height, self.block_time)
         self.blocks.append(block)
         self.block_time += 1
         prepare_block(block)
@@ -135,8 +134,7 @@ class AssumeValidTest(BitcoinTestFramework):
         # Bury the block 100 deep so the coinbase output is spendable
         for i in range(100):
             block = create_block(
-                self.tip, create_coinbase(height), self.block_time)
-            block.nHeight = height
+                self.tip, create_coinbase(height), height, self.block_time)
             prepare_block(block)
             self.blocks.append(block)
             self.tip = block.sha256
@@ -153,8 +151,7 @@ class AssumeValidTest(BitcoinTestFramework):
         tx.rehash()
 
         block102 = create_block(
-            self.tip, create_coinbase(height), self.block_time)
-        block102.nHeight = height
+            self.tip, create_coinbase(height), height, self.block_time)
         self.block_time += 1
         block102.vtx.extend([tx])
         prepare_block(block102)
@@ -166,8 +163,7 @@ class AssumeValidTest(BitcoinTestFramework):
         # Bury the assumed valid block 3400 deep
         for i in range(3700):
             block = create_block(
-                self.tip, create_coinbase(height), self.block_time)
-            block.nHeight = height
+                self.tip, create_coinbase(height), height, self.block_time)
             prepare_block(block)
             self.blocks.append(block)
             self.tip = block.sha256
diff --git a/test/functional/feature_bip68_sequence.py b/test/functional/feature_bip68_sequence.py
index 65019b2a5..811b59eeb 100755
--- a/test/functional/feature_bip68_sequence.py
+++ b/test/functional/feature_bip68_sequence.py
@@ -396,8 +396,7 @@ class BIP68Test(BitcoinTestFramework):
             self.nodes[0].getblockcount() - 1), 16)
         height = self.nodes[0].getblockcount()
         for i in range(2):
-            block = create_block(tip, create_coinbase(height), cur_time)
-            block.nHeight = height
+            block = create_block(tip, create_coinbase(height), height, cur_time)
             prepare_block(block)
             tip = block.sha256
             height += 1
diff --git a/test/functional/feature_block.py b/test/functional/feature_block.py
index 36abbfae1..c70b6f008 100755
--- a/test/functional/feature_block.py
+++ b/test/functional/feature_block.py
@@ -12,6 +12,7 @@ from test_framework.blocktools import (
     create_block,
     create_coinbase,
     create_tx_with_script,
+    prepare_block,
     make_conform_to_ctor,
 )
 from test_framework.cdefs import LEGACY_MAX_BLOCK_SIZE
@@ -93,12 +94,10 @@ class FullBlockTest(BitcoinTestFramework):
         self.genesis_hash = int(self.nodes[0].getbestblockhash(), 16)
         self.block_heights[self.genesis_hash] = 0
         self.spendable_outputs = []
-
         # Create a new block
         b_dup_cb = self.next_block('dup_cb')
         b_dup_cb.vtx[0].vin[0].scriptSig = DUPLICATE_COINBASE_SCRIPT_SIG
         b_dup_cb.vtx[0].rehash()
-        duplicate_tx = b_dup_cb.vtx[0]
         b_dup_cb = self.update_block('dup_cb', [])
         self.send_blocks([b_dup_cb])
 
@@ -117,7 +116,6 @@ class FullBlockTest(BitcoinTestFramework):
             blocks.append(self.next_block("maturitybuffer.{}".format(i)))
             self.save_spendable_output()
         self.send_blocks(blocks)
-
         # collect spendable outputs now to avoid cluttering the code later on
         out = []
         for i in range(NUM_OUTPUTS_TO_COLLECT):
@@ -128,12 +126,11 @@ class FullBlockTest(BitcoinTestFramework):
         #     genesis -> b1 (0) -> b2 (1)
         b1 = self.next_block(1, spend=out[0])
         self.save_spendable_output()
-
+        self.send_blocks([b1], timeout=4)
         b2 = self.next_block(2, spend=out[1])
         self.save_spendable_output()
 
         self.send_blocks([b1, b2], timeout=4)
-
         # Select a txn with an output eligible for spending. This won't actually be spent,
         # since we're testing submission of a series of blocks with invalid
         # txns.
@@ -281,7 +278,6 @@ class FullBlockTest(BitcoinTestFramework):
         self.move_tip(13)
         b18 = self.next_block(18, spend=txout_b3)
         self.send_blocks([b18], False)
-
         b19 = self.next_block(19, spend=out[6])
         self.send_blocks([b19], success=False,
                          reject_reason='bad-txns-inputs-missingorspent', reconnect=True)
@@ -329,7 +325,7 @@ class FullBlockTest(BitcoinTestFramework):
         script_length = LEGACY_MAX_BLOCK_SIZE - len(b23.serialize()) - 69
         script_output = CScript([b'\x00' * script_length])
         tx.vout.append(CTxOut(0, script_output))
-        tx.vin.append(CTxIn(COutPoint(b23.vtx[1].sha256, 0)))
+        tx.vin.append(CTxIn(COutPoint(b23.vtx[1].txid, 0)))
         b23 = self.update_block(23, [tx])
         # Make sure the math above worked out to produce a max-sized block
         assert_equal(len(b23.serialize()), LEGACY_MAX_BLOCK_SIZE)
@@ -454,6 +450,9 @@ class FullBlockTest(BitcoinTestFramework):
         b44.nBits = 0x207fffff
         b44.vtx.append(coinbase)
         b44.hashMerkleRoot = b44.calc_merkle_root()
+        b44.rehash_extended_metadata()
+        b44.nHeight = height
+        b44.update_size()
         b44.solve()
         self.tip = b44
         self.block_heights[b44.sha256] = height
@@ -468,6 +467,9 @@ class FullBlockTest(BitcoinTestFramework):
         b45.nBits = 0x207fffff
         b45.vtx.append(non_coinbase)
         b45.hashMerkleRoot = b45.calc_merkle_root()
+        b45.rehash_extended_metadata()
+        b45.nHeight = self.block_heights[self.tip.sha256] + 1
+        b45.update_size()
         b45.calc_sha256()
         b45.solve()
         self.block_heights[b45.sha256] = self.block_heights[
@@ -485,6 +487,9 @@ class FullBlockTest(BitcoinTestFramework):
         b46.nBits = 0x207fffff
         b46.vtx = []
         b46.hashMerkleRoot = 0
+        b46.rehash_extended_metadata()
+        b46.nHeight = self.block_heights[self.tip.sha256] + 1
+        b46.update_size()
         b46.solve()
         self.block_heights[b46.sha256] = self.block_heights[b44.sha256] + 1
         self.tip = b46
@@ -604,7 +609,7 @@ class FullBlockTest(BitcoinTestFramework):
         #  Result:  OK.
         #
         #  b56p2 copies b57p2 but adds both tx3 and tx4.  The purpose of the test is to make sure the code catches
-        #  duplicate txns that are not next to one another with the "bad-txns-duplicate" error (which indicates
+        #  duplicate txns that are not next to one another with the "tx-duplicate" error (which indicates
         #  that the error was caught early, avoiding a DOS vulnerability.)
 
         # b57 - a good block with 2 txs, don't submit until end
@@ -621,10 +626,9 @@ class FullBlockTest(BitcoinTestFramework):
         b56 = copy.deepcopy(b57)
         self.blocks[56] = b56
         assert_equal(len(b56.vtx), 3)
-        b56 = self.update_block(56, [b57.vtx[2]])
-        assert_equal(b56.hash, b57.hash)
+        b56 = self.update_block(56, [b57.vtx[2]], height=self.block_heights[self.tip.sha256] + 1)
         self.send_blocks([b56], success=False,
-                         reject_reason='bad-txns-duplicate', reconnect=True)
+                         reject_reason='tx-duplicate', reconnect=True)
 
         # b57p2 - a good block with 6 tx'es, don't submit until end
         self.move_tip(55)
@@ -636,18 +640,6 @@ class FullBlockTest(BitcoinTestFramework):
         tx4 = self.create_tx(tx3, 0, 1)
         b57p2 = self.update_block("57p2", [tx, tx1, tx2, tx3, tx4])
 
-        # b56p2 - copy b57p2, duplicate two non-consecutive tx's
-        self.log.info(
-            "Reject a block with two duplicate transactions in the Merkle Tree (but with a valid Merkle Root)")
-        self.move_tip(55)
-        b56p2 = copy.deepcopy(b57p2)
-        self.blocks["b56p2"] = b56p2
-        assert_equal(len(b56p2.vtx), 6)
-        b56p2 = self.update_block("b56p2", b56p2.vtx[4:6], reorder=False)
-        assert_equal(b56p2.hash, b57p2.hash)
-        self.send_blocks([b56p2], success=False,
-                         reject_reason='bad-txns-duplicate', reconnect=True)
-
         self.move_tip("57p2")
         self.send_blocks([b57p2], True)
 
@@ -670,10 +662,10 @@ class FullBlockTest(BitcoinTestFramework):
         tx = CTransaction()
         assert(len(out[17].vout) < 42)
         tx.vin.append(
-            CTxIn(COutPoint(out[17].sha256, 42), CScript([OP_TRUE]), 0xffffffff))
+            CTxIn(COutPoint(out[17].txid, 42), CScript([OP_TRUE]), 0xffffffff))
         tx.vout.append(CTxOut(0, b""))
         pad_tx(tx)
-        tx.calc_sha256()
+        tx.rehash()
         b58 = self.update_block(58, [tx])
         self.send_blocks([b58], success=False,
                          reject_reason='bad-txns-inputs-missingorspent', reconnect=True)
@@ -683,68 +675,18 @@ class FullBlockTest(BitcoinTestFramework):
             "Reject a block with a transaction with outputs > inputs")
         self.move_tip(57)
         b59 = self.next_block(59)
-        tx = self.create_and_sign_transaction(out[17], 51 * COIN)
+        tx = self.create_and_sign_transaction(out[17], 261 * COIN)
         b59 = self.update_block(59, [tx])
         self.send_blocks([b59], success=False,
                          reject_reason='bad-txns-in-belowout', reconnect=True)
 
         # reset to good chain
         self.move_tip(57)
+
         b60 = self.next_block(60)
         self.send_blocks([b60], True)
         self.save_spendable_output()
-
-        # Test BIP30 (reject duplicate)
-        #
-        # -> b39 (11) -> b42 (12) -> b43 (13) -> b53 (14) -> b55 (15) -> b57 (16) -> b60 ()
-        #                                                                                  \-> b61 ()
-        #
-        # Blocks are not allowed to contain a transaction whose id matches that of an earlier,
-        # not-fully-spent transaction in the same chain. To test, make identical coinbases;
-        # the second one should be rejected. See also CVE-2012-1909.
-        #
-        self.log.info(
-            "Reject a block with a transaction with a duplicate hash of a previous transaction (BIP30)")
-        self.move_tip(60)
-        b61 = self.next_block(61)
-        b61.vtx[0].vin[0].scriptSig = DUPLICATE_COINBASE_SCRIPT_SIG
-        b61.vtx[0].rehash()
-        b61 = self.update_block(61, [])
-        assert_equal(duplicate_tx.serialize(), b61.vtx[0].serialize())
-        self.send_blocks([b61], success=False,
-                         reject_reason='bad-txns-BIP30', reconnect=True)
-
-        # Test BIP30 (allow duplicate if spent)
-        #
-        # -> b57 (16) -> b60 ()
-        #            \-> b_spend_dup_cb (b_dup_cb) -> b_dup_2 ()
-        #
-        self.move_tip(57)
-        b_spend_dup_cb = self.next_block('spend_dup_cb')
-        tx = CTransaction()
-        tx.vin.append(CTxIn(COutPoint(duplicate_tx.sha256, 0)))
-        tx.vout.append(CTxOut(0, CScript([OP_TRUE])))
-        self.sign_tx(tx, duplicate_tx)
-        tx.rehash()
-        b_spend_dup_cb = self.update_block('spend_dup_cb', [tx])
-
-        b_dup_2 = self.next_block('dup_2')
-        b_dup_2.vtx[0].vin[0].scriptSig = DUPLICATE_COINBASE_SCRIPT_SIG
-        b_dup_2.vtx[0].rehash()
-        b_dup_2 = self.update_block('dup_2', [])
-        assert_equal(duplicate_tx.serialize(), b_dup_2.vtx[0].serialize())
-        assert_equal(
-            self.nodes[0].gettxout(
-                txid=duplicate_tx.hash,
-                n=0)['confirmations'],
-            119)
-        self.send_blocks([b_spend_dup_cb, b_dup_2], success=True)
-        # The duplicate has less confirmations
-        assert_equal(
-            self.nodes[0].gettxout(
-                txid=duplicate_tx.hash,
-                n=0)['confirmations'],
-            1)
+        
 
         # Test tx.isFinal is properly rejected (not an exhaustive tx.isFinal test, that should be in data-driven transaction tests)
         #
@@ -753,15 +695,14 @@ class FullBlockTest(BitcoinTestFramework):
         #
         self.log.info(
             "Reject a block with a transaction with a nonfinal locktime")
-        self.move_tip('dup_2')
         b62 = self.next_block(62)
         tx = CTransaction()
         tx.nLockTime = 0xffffffff  # this locktime is non-final
         # don't set nSequence
-        tx.vin.append(CTxIn(COutPoint(out[18].sha256, 0)))
+        tx.vin.append(CTxIn(COutPoint(out[18].txid, 0)))
         tx.vout.append(CTxOut(0, CScript([OP_TRUE])))
         assert tx.vin[0].nSequence < 0xffffffff
-        tx.calc_sha256()
+        tx.rehash()
         b62 = self.update_block(62, [tx])
         self.send_blocks(
             [b62],
@@ -776,7 +717,7 @@ class FullBlockTest(BitcoinTestFramework):
         #
         self.log.info(
             "Reject a block with a coinbase transaction with a nonfinal locktime")
-        self.move_tip('dup_2')
+        self.move_tip(57)
         b63 = self.next_block(63)
         b63.vtx[0].nLockTime = 0xffffffff
         b63.vtx[0].vin[0].nSequence = 0xDEADBEEF
@@ -804,7 +745,7 @@ class FullBlockTest(BitcoinTestFramework):
         #
         self.log.info(
             "Accept a valid block even if a bloated version of the block has previously been sent")
-        self.move_tip('dup_2')
+        self.move_tip(60)
         regular_block = self.next_block("64a", spend=out[18])
 
         # make it a "broken_block," with non-canonical serialization
@@ -819,10 +760,10 @@ class FullBlockTest(BitcoinTestFramework):
             len(b64a.normal_serialize()) - 69
         script_output = CScript([b'\x00' * script_length])
         tx.vout.append(CTxOut(0, script_output))
-        tx.vin.append(CTxIn(COutPoint(b64a.vtx[1].sha256, 0)))
+        tx.vin.append(CTxIn(COutPoint(b64a.vtx[1].txid, 0)))
         b64a = self.update_block("64a", [tx])
-        assert_equal(len(b64a.serialize()), LEGACY_MAX_BLOCK_SIZE + 8)
-        self.send_blocks([b64a], success=False,
+        assert_equal(len(b64a.serialize()), LEGACY_MAX_BLOCK_SIZE + 7)
+        self.send_blocks([b64a], success=False, force_send=True,
                          reject_reason='non-canonical ReadCompactSize()')
 
         # lotusd doesn't disconnect us for sending a bloated block, but if we subsequently
@@ -832,13 +773,13 @@ class FullBlockTest(BitcoinTestFramework):
         node.disconnect_p2ps()
         self.reconnect_p2p()
 
-        self.move_tip('dup_2')
+        self.move_tip(60)
         b64 = CBlock(b64a)
         b64.vtx = copy.deepcopy(b64a.vtx)
         assert_equal(b64.hash, b64a.hash)
         assert_equal(len(b64.serialize()), LEGACY_MAX_BLOCK_SIZE)
         self.blocks[64] = b64
-        b64 = self.update_block(64, [])
+        b64 = self.update_block(64, [], height=self.block_heights[self.tip.sha256] + 1)
         self.send_blocks([b64], True)
         self.save_spendable_output()
 
@@ -866,7 +807,7 @@ class FullBlockTest(BitcoinTestFramework):
             "Reject a block with a transaction double spending a transaction created in the same block")
         self.move_tip(65)
         b67 = self.next_block(67)
-        tx1 = self.create_and_sign_transaction(out[20], out[20].vout[0].nValue)
+        tx1 = self.create_and_sign_transaction(out[20], out[20].vout[1].nValue)
         tx2 = self.create_and_sign_transaction(tx1, 1)
         tx3 = self.create_and_sign_transaction(tx1, 2)
         b67 = self.update_block(67, [tx1, tx2, tx3])
@@ -890,7 +831,7 @@ class FullBlockTest(BitcoinTestFramework):
         self.move_tip(65)
         b68 = self.next_block(68, additional_coinbase_value=10)
         tx = self.create_and_sign_transaction(
-            out[20], out[20].vout[0].nValue - 9)
+            out[20], out[20].vout[1].nValue - 9)
         b68 = self.update_block(68, [tx])
         self.send_blocks([b68], success=False,
                          reject_reason='bad-cb-amount', reconnect=True)
@@ -898,9 +839,9 @@ class FullBlockTest(BitcoinTestFramework):
         self.log.info(
             "Accept a block claiming the correct subsidy in the coinbase transaction")
         self.move_tip(65)
-        b69 = self.next_block(69, additional_coinbase_value=10)
+        b69 = self.next_block(69, additional_coinbase_value=5)
         tx = self.create_and_sign_transaction(
-            out[20], out[20].vout[0].nValue - 10)
+            out[20], out[20].vout[1].nValue - 10)
         self.update_block(69, [tx])
         self.send_blocks([b69], True)
         self.save_spendable_output()
@@ -915,10 +856,10 @@ class FullBlockTest(BitcoinTestFramework):
         self.move_tip(69)
         b70 = self.next_block(70, spend=out[21])
         bogus_tx = CTransaction()
-        bogus_tx.sha256 = uint256_from_str(
+        bogus_tx.txid = uint256_from_str(
             b"23c70ed7c0506e9178fc1a987f40a33946d4ad4c962b5ae3a52546da53af0c5c")
         tx = CTransaction()
-        tx.vin.append(CTxIn(COutPoint(bogus_tx.sha256, 0), b"", 0xffffffff))
+        tx.vin.append(CTxIn(COutPoint(bogus_tx.txid, 0), b"", 0xffffffff))
         tx.vout.append(CTxOut(1, b""))
         pad_tx(tx)
         b70 = self.update_block(70, [tx])
@@ -953,7 +894,7 @@ class FullBlockTest(BitcoinTestFramework):
 
         self.move_tip(71)
         self.send_blocks([b71], success=False,
-                         reject_reason='bad-txns-duplicate', reconnect=True)
+                         reject_reason='bad-txnmrklroot', reconnect=True)
 
         self.move_tip(72)
         self.send_blocks([b72], True)
@@ -1026,15 +967,15 @@ class FullBlockTest(BitcoinTestFramework):
         # mempool
         mempool = self.nodes[0].getrawmempool()
         assert_equal(len(mempool), 2)
-        assert tx78.hash in mempool
-        assert tx79.hash in mempool
+        assert tx78.txid_hex in mempool
+        assert tx79.txid_hex in mempool
 
         # Test invalid opcodes in dead execution paths.
         #
         #  -> b81 (26) -> b82 (27) -> b83 (28)
         #
         self.log.info(
-            "Accept a block with invalid opcodes in dead execution paths")
+            "Reject a block with invalid opcodes in dead execution paths")
         b83 = self.next_block(83)
         op_codes = [OP_IF, OP_INVALIDOPCODE, OP_ELSE, OP_TRUE, OP_ENDIF]
         script = CScript(op_codes)
@@ -1046,8 +987,7 @@ class FullBlockTest(BitcoinTestFramework):
         tx2.rehash()
 
         b83 = self.update_block(83, [tx1, tx2])
-        self.send_blocks([b83], True)
-        self.save_spendable_output()
+        self.send_blocks([b83], success=False, reject_reason='blk-bad-inputs', reconnect=True)
 
         # Reorg on/off blocks that have OP_RETURN in them (and try to spend them)
         #
@@ -1055,15 +995,16 @@ class FullBlockTest(BitcoinTestFramework):
         #                                    \-> b85 (29) -> b86 (30)            \-> b89a (32)
         #
         self.log.info("Test re-orging blocks with OP_RETURN in them")
+        self.move_tip(82)
         b84 = self.next_block(84)
-        tx1 = self.create_tx(out[29], 0, 0, CScript([OP_RETURN]))
+        tx1 = self.create_tx(out[28], 1, 0, CScript([OP_RETURN]))
         vout_offset = len(tx1.vout)
         tx1.vout.append(CTxOut(0, CScript([OP_TRUE])))
         tx1.vout.append(CTxOut(0, CScript([OP_TRUE])))
         tx1.vout.append(CTxOut(0, CScript([OP_TRUE])))
         tx1.vout.append(CTxOut(0, CScript([OP_TRUE])))
-        tx1.calc_sha256()
-        self.sign_tx(tx1, out[29])
+        tx1.rehash()
+        self.sign_tx(tx1, out[28])
         tx1.rehash()
         tx2 = self.create_tx(tx1, vout_offset, 0, CScript([OP_RETURN]))
         tx2.vout.append(CTxOut(0, CScript([OP_RETURN])))
@@ -1075,26 +1016,27 @@ class FullBlockTest(BitcoinTestFramework):
 
         b84 = self.update_block(84, [tx1, tx2, tx3, tx4, tx5])
         self.send_blocks([b84], True)
+
         self.save_spendable_output()
 
-        self.move_tip(83)
-        b85 = self.next_block(85, spend=out[29])
+        self.move_tip(82)
+        b85 = self.next_block(85, spend=out[28])
         self.send_blocks([b85], False)  # other chain is same length
 
-        b86 = self.next_block(86, spend=out[30])
+        b86 = self.next_block(86, spend=out[29])
         self.send_blocks([b86], True)
 
         self.move_tip(84)
-        b87 = self.next_block(87, spend=out[30])
+        b87 = self.next_block(87, spend=out[29])
         self.send_blocks([b87], False)  # other chain is same length
         self.save_spendable_output()
 
-        b88 = self.next_block(88, spend=out[31])
+        b88 = self.next_block(88, spend=out[30])
         self.send_blocks([b88], True)
         self.save_spendable_output()
 
         # trying to spend the OP_RETURN output is rejected
-        b89a = self.next_block("89a", spend=out[32])
+        b89a = self.next_block("89a", spend=out[31])
         tx = self.create_tx(tx1, 0, 0, CScript([OP_TRUE]))
         b89a = self.update_block("89a", [tx])
         self.send_blocks([b89a], success=False,
@@ -1103,59 +1045,71 @@ class FullBlockTest(BitcoinTestFramework):
         self.log.info(
             "Test a re-org of one week's worth of blocks (1088 blocks)")
 
+        self.log.info("Reject a block with an invalid block header version")
         self.move_tip(88)
+
+        # Due to previous tests that were removed, add in a spacer to mature
+        # some of the spent coinbases below.
+        self.send_blocks([self.next_block('maturity.spacer.1')], success=True)
+        self.save_spendable_output()
+
         LARGE_REORG_SIZE = 1088
         blocks = []
         spend = out[32]
+        print(spend)
         for i in range(89, LARGE_REORG_SIZE + 89):
-            b = self.next_block(i, spend, version=4)
+            b = self.next_block(i, spend)
             tx = CTransaction()
             script_length = LEGACY_MAX_BLOCK_SIZE - len(b.serialize()) - 69
             script_output = CScript([b'\x00' * script_length])
             tx.vout.append(CTxOut(0, script_output))
-            tx.vin.append(CTxIn(COutPoint(b.vtx[1].sha256, 0)))
+            tx.vin.append(CTxIn(COutPoint(b.vtx[1].txid, 0)))
             b = self.update_block(i, [tx])
             assert_equal(len(b.serialize()), LEGACY_MAX_BLOCK_SIZE)
             blocks.append(b)
             self.save_spendable_output()
             spend = self.get_spendable_output()
 
-        self.send_blocks(blocks, True, timeout=1920)
+        self.send_blocks(blocks, True, timeout=360)
         chain1_tip = i
 
         # now create alt chain of same length
         self.move_tip(88)
+        # Won't succeed getbestblockhash because it doesn't trigger a fork.
+        self.send_blocks([self.next_block('maturity.spacer.2')], success=False)
+        self.save_spendable_output()
+
         blocks2 = []
         for i in range(89, LARGE_REORG_SIZE + 89):
-            blocks2.append(self.next_block("alt" + str(i), version=4))
+            blocks2.append(self.next_block("alt" + str(i)))
         self.send_blocks(blocks2, False, force_send=True)
 
         # extend alt chain to trigger re-org
-        block = self.next_block("alt" + str(chain1_tip + 1), version=4)
-        self.send_blocks([block], True, timeout=1920)
+        block = self.next_block("alt" + str(chain1_tip + 1))
+        self.send_blocks([block], True, timeout=360)
 
         # ... and re-org back to the first chain
         self.move_tip(chain1_tip)
-        block = self.next_block(chain1_tip + 1, version=4)
+        block = self.next_block(chain1_tip + 1)
         self.send_blocks([block], False, force_send=True)
-        block = self.next_block(chain1_tip + 2, version=4)
-        self.send_blocks([block], True, timeout=1920)
+        block = self.next_block(chain1_tip + 2)
+        self.send_blocks([block], True, timeout=360)
 
         self.log.info("Reject a block with an invalid block header version")
-        b_v1 = self.next_block('b_v1', version=1)
+        b_v4 = self.next_block('b_v4', version=4)
         self.send_blocks(
-            [b_v1],
+            [b_v4],
             success=False,
             force_send=True,
-            reject_reason='bad-version(0x00000001)',
+            reject_reason='bad-blk-version',
             reconnect=True)
 
+        self.log.info("Reject a block with an invalid block height in coinbase")
         self.move_tip(chain1_tip + 2)
-        b_cb34 = self.next_block('b_cb34', version=4)
-        b_cb34.vtx[0].vin[0].scriptSig = b_cb34.vtx[0].vin[0].scriptSig[:-1]
-        b_cb34.vtx[0].rehash()
-        b_cb34.hashMerkleRoot = b_cb34.calc_merkle_root()
-        b_cb34.solve()
+        b_cb34 = self.next_block('b_cb34', version=1)
+        b_cb34.vtx[0] = create_coinbase(height=b_cb34.nHeight+1)
+        b_cb34 = self.update_block('b_cb34', [])
+        self.log.info(b_cb34.hash)
         self.send_blocks(
             [b_cb34],
             success=False,
@@ -1178,18 +1132,24 @@ class FullBlockTest(BitcoinTestFramework):
     # this signs input 0 in tx, which is assumed to be spending output n in
     # spend_tx
     def sign_tx(self, tx, spend_tx):
-        scriptPubKey = bytearray(spend_tx.vout[0].scriptPubKey)
+        n = 0
+        if spend_tx.is_coinbase():
+            n = 1
+        scriptPubKey = bytearray(spend_tx.vout[n].scriptPubKey)
         if (scriptPubKey[0] == OP_TRUE):  # an anyone-can-spend
             tx.vin[0].scriptSig = CScript()
             return
         sighash = SignatureHashForkId(
-            spend_tx.vout[0].scriptPubKey, tx, 0, SIGHASH_ALL | SIGHASH_FORKID, spend_tx.vout[0].nValue)
+            spend_tx.vout[n].scriptPubKey, tx, 0, SIGHASH_ALL | SIGHASH_FORKID, spend_tx.vout[n].nValue)
         tx.vin[0].scriptSig = CScript(
             [self.coinbase_key.sign_ecdsa(sighash) + bytes(bytearray([SIGHASH_ALL | SIGHASH_FORKID]))])
 
     def create_and_sign_transaction(
             self, spend_tx, value, script=CScript([OP_TRUE])):
-        tx = self.create_tx(spend_tx, 0, value, script)
+        vin = 0
+        if spend_tx.is_coinbase():
+            vin = 1
+        tx = self.create_tx(spend_tx, vin, value, script)
         self.sign_tx(tx, spend_tx)
         tx.rehash()
         return tx
@@ -1205,30 +1165,34 @@ class FullBlockTest(BitcoinTestFramework):
         # First create the coinbase
         height = self.block_heights[base_block_hash] + 1
         coinbase = create_coinbase(height, self.coinbase_pubkey)
-        coinbase.vout[0].nValue += additional_coinbase_value
+        coinbase.vout[1].nValue += additional_coinbase_value
         coinbase.rehash()
         if spend is None:
             block = create_block(
                 base_block_hash,
                 coinbase,
+                height,
                 block_time,
                 version=version)
         else:
+            vout = 0
+            if spend.is_coinbase():
+                vout = 1
             # all but one satoshi to fees
-            coinbase.vout[0].nValue += spend.vout[0].nValue - 1
+            coinbase.vout[0].nValue += (spend.vout[vout].nValue - 1) // 2
             coinbase.rehash()
             block = create_block(
                 base_block_hash,
                 coinbase,
+                height,
                 block_time,
                 version=version)
             # spend 1 satoshi
-            tx = self.create_tx(spend, 0, 1, script)
+            tx = self.create_tx(spend, vout, 1, script)
             self.sign_tx(tx, spend)
             self.add_transactions_to_block(block, [tx])
-            block.hashMerkleRoot = block.calc_merkle_root()
         # Block is created. Find a valid nonce.
-        block.solve()
+        prepare_block(block)
         self.tip = block
         self.block_heights[block.sha256] = height
         assert number not in self.blocks
@@ -1237,13 +1201,12 @@ class FullBlockTest(BitcoinTestFramework):
 
     # save the current tip so it can be spent by a later block
     def save_spendable_output(self):
-        self.log.debug("saving spendable output {}".format(self.tip.vtx[0]))
+        self.log.debug("saving spendable outputs")
         self.spendable_outputs.append(self.tip)
 
     # get an output that we previously marked as spendable
     def get_spendable_output(self):
-        self.log.debug("getting spendable output {}".format(
-            self.spendable_outputs[0].vtx[0]))
+        self.log.debug("getting spendable output")
         return self.spendable_outputs.pop(0).vtx[0]
 
     # move the tip back to a previous block
@@ -1251,19 +1214,22 @@ class FullBlockTest(BitcoinTestFramework):
         self.tip = self.blocks[number]
 
     # adds transactions to the block and updates state
-    def update_block(self, block_number, new_transactions, reorder=True):
+    def update_block(self, block_number, new_transactions, reorder=True, height=None):
         block = self.blocks[block_number]
         self.add_transactions_to_block(block, new_transactions)
         old_sha256 = block.sha256
         if reorder:
             make_conform_to_ctor(block)
+        block.rehash_extended_metadata()
+        block.update_size()
         block.hashMerkleRoot = block.calc_merkle_root()
+        if height:
+            block.nHeight = height
         block.solve()
         # Update the internal state just like in next_block
         self.tip = block
         if block.sha256 != old_sha256:
             self.block_heights[block.sha256] = self.block_heights[old_sha256]
-            del self.block_heights[old_sha256]
         self.blocks[block_number] = block
         return block
 
diff --git a/test/functional/feature_cltv.py b/test/functional/feature_cltv.py
index bbd1468a9..bad4bf675 100755
--- a/test/functional/feature_cltv.py
+++ b/test/functional/feature_cltv.py
@@ -117,8 +117,7 @@ class BIP65Test(BitcoinTestFramework):
         tip = self.nodes[0].getbestblockhash()
         block_time = self.nodes[0].getblockheader(tip)['mediantime'] + 1
         block = create_block(int(tip, 16), create_coinbase(
-            CLTV_HEIGHT - 1), block_time)
-        block.nHeight = CLTV_HEIGHT - 1
+            CLTV_HEIGHT - 1), CLTV_HEIGHT,  block_time)
         block.vtx.append(fundtx)
         # include the -1 CLTV in block
         block.vtx.append(spendtx)
@@ -130,8 +129,7 @@ class BIP65Test(BitcoinTestFramework):
         
         # Create valid block to get over the threshold for the version enforcement
         block = create_block(int(tip, 16), create_coinbase(
-            CLTV_HEIGHT - 1), block_time)
-        block.nHeight = CLTV_HEIGHT - 1
+            CLTV_HEIGHT - 1), CLTV_HEIGHT - 1, block_time)
         prepare_block(block)
         self.nodes[0].p2p.send_and_ping(msg_block(block))
 
@@ -139,8 +137,7 @@ class BIP65Test(BitcoinTestFramework):
         block_time += 1
         self.log.info(
             "Test that invalid-according-to-cltv transactions cannot appear in a block")
-        block = create_block(tip, create_coinbase(CLTV_HEIGHT), block_time)
-        block.nHeight = CLTV_HEIGHT
+        block = create_block(tip, create_coinbase(CLTV_HEIGHT), CLTV_HEIGHT, block_time)
 
         fundtx = create_transaction(self.nodes[0], self.coinbase_txids[1],
                                     self.nodeaddress, amount=SUBSIDY - Decimal('1'), vout=1)
@@ -178,8 +175,7 @@ class BIP65Test(BitcoinTestFramework):
         tip = block.hash
         block_time += 1
         block = create_block(
-            block.sha256, create_coinbase(CLTV_HEIGHT + 1), block_time)
-        block.nHeight = CLTV_HEIGHT + 1
+            block.sha256, create_coinbase(CLTV_HEIGHT + 1), CLTV_HEIGHT + 1, block_time)
         block.vtx.append(spendtx)
         prepare_block(block)
 
diff --git a/test/functional/feature_csv_activation.py b/test/functional/feature_csv_activation.py
index 54c4ca6dc..a42883eb3 100755
--- a/test/functional/feature_csv_activation.py
+++ b/test/functional/feature_csv_activation.py
@@ -201,9 +201,8 @@ class BIP68_112_113Test(BitcoinTestFramework):
 
     def create_test_block(self, txs):
         block = create_block(self.tip, create_coinbase(
-            self.tipheight + 1), self.last_block_time + 600)
+            self.tipheight + 1), self.tipheight + 1, self.last_block_time + 600)
         block.vtx.extend(txs)
-        block.nHeight = self.tipheight + 1
         prepare_block(block)
         return block
 
diff --git a/test/functional/feature_dersig.py b/test/functional/feature_dersig.py
index 88175aafe..516718eb6 100755
--- a/test/functional/feature_dersig.py
+++ b/test/functional/feature_dersig.py
@@ -59,8 +59,7 @@ class BIP66Test(BitcoinTestFramework):
         tip = self.nodes[0].getbestblockhash()
         block_time = self.nodes[0].getblockheader(tip)['mediantime'] + 1
         block = create_block(
-            int(tip, 16), create_coinbase(DERSIG_HEIGHT), block_time)
-        block.nHeight = DERSIG_HEIGHT
+            int(tip, 16), create_coinbase(DERSIG_HEIGHT), DERSIG_HEIGHT, block_time)
         spendtx = create_transaction(self.nodes[0], self.coinbase_txids[1],
                                      self.nodeaddress, amount=1.0, vout=1)
         unDERify(spendtx)
diff --git a/test/functional/feature_pruning.py b/test/functional/feature_pruning.py
index a108fcd19..8c355dd4a 100755
--- a/test/functional/feature_pruning.py
+++ b/test/functional/feature_pruning.py
@@ -153,7 +153,7 @@ class PruneTest(BitcoinTestFramework):
 
         # Wait for blk00000.dat to be pruned
         wait_until(lambda: not os.path.isfile(
-            os.path.join(self.prunedir, "blk00000.dat")), timeout=30)
+            os.path.join(self.prunedir, "blk00000.dat")), timeout=120)
 
         self.log.info("Success")
         usage = calc_usage(self.prunedir)
diff --git a/test/functional/logos_enforce_standard_consensus.py b/test/functional/logos_enforce_standard_consensus.py
index 1efa1c193..bf9d331a5 100755
--- a/test/functional/logos_enforce_standard_consensus.py
+++ b/test/functional/logos_enforce_standard_consensus.py
@@ -76,8 +76,7 @@ class EnforceStandardConsensusTest(BitcoinTestFramework):
             coinbase.vout[1].scriptPubKey = p2sh_script
             coinbase.rehash()
             block = create_block(
-                int(parent_block_header['hash'], 16), coinbase, parent_block_header['time'] + 1)
-            block.nHeight = height
+                int(parent_block_header['hash'], 16), coinbase, height, parent_block_header['time'] + 1)
             return block
 
         # make a few non-standard txs
diff --git a/test/functional/logos_feature_int63.py b/test/functional/logos_feature_int63.py
index 0f8ff9b11..32329c1df 100755
--- a/test/functional/logos_feature_int63.py
+++ b/test/functional/logos_feature_int63.py
@@ -86,8 +86,7 @@ class Int63Test(BitcoinTestFramework):
             coinbase.vout[1].scriptPubKey = p2sh_script
             coinbase.rehash()
             block = create_block(
-                int(parent_block_header['hash'], 16), coinbase, parent_block_header['time'] + 1)
-            block.nHeight = height
+                int(parent_block_header['hash'], 16), coinbase, height, parent_block_header['time'] + 1)
             return block
 
         interesting_numbers = [
diff --git a/test/functional/logos_feature_minerfund.py b/test/functional/logos_feature_minerfund.py
index 9155a9f7a..31378dd06 100755
--- a/test/functional/logos_feature_minerfund.py
+++ b/test/functional/logos_feature_minerfund.py
@@ -66,8 +66,7 @@ class MinerFundTest(BitcoinTestFramework):
         coinbase.vout[1].nValue = int(SUBSIDY * COIN)
         coinbase.rehash()
         block = create_block(int(best_block_hash, 16),
-                             coinbase, block_time + 1)
-        block.nHeight = block_height
+                             coinbase, block_height, block_time + 1)
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
 
@@ -83,8 +82,7 @@ class MinerFundTest(BitcoinTestFramework):
         coinbase.vout[2], coinbase.vout[4] = coinbase.vout[4], coinbase.vout[2]
         coinbase.rehash()
         block = create_block(int(best_block_hash, 16),
-                             coinbase, block_time + 1)
-        block.nHeight = block_height
+                             coinbase, block_height, block_time + 1)
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
 
@@ -92,9 +90,8 @@ class MinerFundTest(BitcoinTestFramework):
         coinbase.vout[2], coinbase.vout[4] = coinbase.vout[4], coinbase.vout[2]
         coinbase.rehash()
         block = create_block(int(best_block_hash, 16),
-                             coinbase, block_time + 1)
+                             coinbase, block_height, block_time + 1)
 
-        block.nHeight = block_height
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), None)
 
@@ -119,8 +116,7 @@ class MinerFundTest(BitcoinTestFramework):
                                         CScript(bytes.fromhex(output['scriptPubKey']))))
         coinbase.rehash()
         block = create_block(int(best_block_hash, 16),
-                             coinbase, best_block_header['time'] + 1)
-        block.nHeight = block_height
+                             coinbase, block_height, best_block_header['time'] + 1)
         block.vtx.append(tx)
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), None)
diff --git a/test/functional/mempool_updatefromblock.py b/test/functional/mempool_updatefromblock.py
index d377b83b8..304adf004 100755
--- a/test/functional/mempool_updatefromblock.py
+++ b/test/functional/mempool_updatefromblock.py
@@ -25,7 +25,7 @@ class MempoolUpdateFromBlockTest(BitcoinTestFramework):
 
     def transaction_graph_test(self, size, n_tx_to_mine=None,
                                start_input_txid='', end_address='',
-                               fee=Decimal(0.00100000)):
+                               fee=Decimal(0.100000)):
         """Create an acyclic tournament (a type of directed graph) of
         transactions and use it for testing.
 
@@ -42,7 +42,6 @@ class MempoolUpdateFromBlockTest(BitcoinTestFramework):
 
         More details: https://en.wikipedia.org/wiki/Tournament_(graph_theory)
         """
-
         if not start_input_txid:
             start_input_txid = self.nodes[0].getblock(
                 self.nodes[0].getblockhash(1))['tx'][0]
@@ -58,9 +57,9 @@ class MempoolUpdateFromBlockTest(BitcoinTestFramework):
             self.log.debug('Preparing transaction #{}...'.format(i))
             # Prepare inputs.
             if i == 0:
-                inputs = [{'txid': start_input_txid, 'vout': 0}]
+                inputs = [{'txid': start_input_txid, 'vout': 1}]
                 inputs_value = self.nodes[0].gettxout(
-                    start_input_txid, 0)['value']
+                    start_input_txid, 1)['value']
             else:
                 inputs = []
                 inputs_value = 0
diff --git a/test/functional/p2p_compactblocks.py b/test/functional/p2p_compactblocks.py
index 3ea6fa480..c41a02f2d 100755
--- a/test/functional/p2p_compactblocks.py
+++ b/test/functional/p2p_compactblocks.py
@@ -150,8 +150,7 @@ class CompactBlocksTest(BitcoinTestFramework):
         tip = node.getbestblockhash()
         mtp = node.getblockheader(tip)['mediantime']
         block = create_block(
-            int(tip, 16), create_coinbase(height), mtp + 1)
-        block.nHeight = height
+            int(tip, 16), create_coinbase(height), height, mtp + 1)
         prepare_block(block)
         return block
 
diff --git a/test/functional/p2p_fingerprint.py b/test/functional/p2p_fingerprint.py
index 8b8ab6b5f..796d785bd 100755
--- a/test/functional/p2p_fingerprint.py
+++ b/test/functional/p2p_fingerprint.py
@@ -36,8 +36,7 @@ class P2PFingerprintTest(BitcoinTestFramework):
         for _ in range(nblocks):
             coinbase = create_coinbase(prev_height + 1)
             block_time = prev_median_time + 1
-            block = create_block(int(prev_hash, 16), coinbase, block_time)
-            block.nHeight = prev_height + 1
+            block = create_block(int(prev_hash, 16), coinbase, prev_height + 1, block_time)
             prepare_block(block)
 
             blocks.append(block)
diff --git a/test/functional/p2p_invalid_block.py b/test/functional/p2p_invalid_block.py
index 760e21cb4..bf52bb243 100755
--- a/test/functional/p2p_invalid_block.py
+++ b/test/functional/p2p_invalid_block.py
@@ -45,8 +45,7 @@ class InvalidBlockRequestTest(BitcoinTestFramework):
         self.log.info("Create a new block with an anyone-can-spend coinbase")
 
         height = 1
-        block = create_block(tip, create_coinbase(height), block_time)
-        block.nHeight = height
+        block = create_block(tip, create_coinbase(height), height, block_time)
         prepare_block(block)
         # Save the coinbase for later
         block1 = block
@@ -70,8 +69,7 @@ class InvalidBlockRequestTest(BitcoinTestFramework):
         # src/consensus/merkle.cpp.
         self.log.info("Test merkle root malleability.")
 
-        block2 = create_block(tip, create_coinbase(height), block_time)
-        block2.nHeight = height
+        block2 = create_block(tip, create_coinbase(height), height, block_time)
         block_time += 1
 
         # b'0x51' is OP_TRUE
@@ -113,8 +111,7 @@ class InvalidBlockRequestTest(BitcoinTestFramework):
 
         self.log.info("Test very broken block.")
 
-        block3 = create_block(tip, create_coinbase(height), block_time)
-        block3.nHeight = height
+        block3 = create_block(tip, create_coinbase(height), height,  block_time)
         block_time += 1
         block3.vtx[0].vout[0].nValue = int(2 * SUBSIDY * COIN)  # Too high!
         block3.vtx[0].rehash()
@@ -139,8 +136,7 @@ class InvalidBlockRequestTest(BitcoinTestFramework):
 
         # Complete testing of CVE-2018-17144, by checking for the inflation bug.
         # Create a block that spends the output of a tx in a previous block.
-        block4 = create_block(tip, create_coinbase(height), block_time)
-        block4.nHeight = height
+        block4 = create_block(tip, create_coinbase(height), height, block_time)
         tx3 = create_tx_with_script(tx2, 0, script_sig=b'\x51',
                                     amount=int(SUBSIDY * COIN))
 
diff --git a/test/functional/p2p_invalid_tx.py b/test/functional/p2p_invalid_tx.py
index cf71b1805..0a033bc66 100755
--- a/test/functional/p2p_invalid_tx.py
+++ b/test/functional/p2p_invalid_tx.py
@@ -73,8 +73,7 @@ class InvalidTxRequestTest(BitcoinTestFramework):
         height = 1
         blocks = []
         for _ in invalid_txs.iter_all_templates():
-            block = create_block(tip, create_coinbase(height), block_time)
-            block.nHeight = height
+            block = create_block(tip, create_coinbase(height), height, block_time)
             prepare_block(block)
             block_time = block.nTime + 1
             height += 1
diff --git a/test/functional/p2p_sendheaders.py b/test/functional/p2p_sendheaders.py
index c789fae59..08d1a598f 100755
--- a/test/functional/p2p_sendheaders.py
+++ b/test/functional/p2p_sendheaders.py
@@ -269,7 +269,7 @@ class SendHeadersTest(BitcoinTestFramework):
             "Verify getheaders with null locator and invalid hashstop does not return headers.")
         height = tip["height"] + 1
         block = create_block(int(tip["hash"], 16), create_coinbase(height),
-            tip["mediantime"] + 1)
+            height, tip["mediantime"] + 1)
         block.nHeight = height
         prepare_block(block)
         test_node.send_header_for_blocks([block])
@@ -314,9 +314,8 @@ class SendHeadersTest(BitcoinTestFramework):
                 last_time = self.nodes[0].getblock(
                     self.nodes[0].getbestblockhash())['time']
                 block_time = last_time + 1
-                new_block = create_block(
-                    tip, create_coinbase(height + 1), block_time)
-                new_block.nHeight = height + 1
+                new_block = create_block(tip, create_coinbase(height + 1),
+                    height + 1, block_time)
                 prepare_block(new_block)
                 test_node.send_header_for_blocks([new_block])
                 test_node.wait_for_getdata([new_block.sha256])
@@ -356,8 +355,7 @@ class SendHeadersTest(BitcoinTestFramework):
                 blocks = []
                 for b in range(i + 1):
                     blocks.append(create_block(
-                        tip, create_coinbase(height), block_time))
-                    blocks[-1].nHeight = height
+                        tip, create_coinbase(height), height, block_time))
                     prepare_block(blocks[-1])
                     tip = blocks[-1].sha256
                     block_time += 1
@@ -480,8 +478,7 @@ class SendHeadersTest(BitcoinTestFramework):
         blocks = []
         for b in range(2):
             blocks.append(create_block(
-                tip, create_coinbase(height), block_time))
-            blocks[-1].nHeight = height
+                tip, create_coinbase(height), height, block_time))
             prepare_block(blocks[-1])
             tip = blocks[-1].sha256
             block_time += 1
@@ -500,8 +497,7 @@ class SendHeadersTest(BitcoinTestFramework):
         blocks = []
         for b in range(3):
             blocks.append(create_block(
-                tip, create_coinbase(height), block_time))
-            blocks[-1].nHeight = height
+                tip, create_coinbase(height), height, block_time))
             prepare_block(blocks[-1])
             tip = blocks[-1].sha256
             block_time += 1
@@ -524,8 +520,7 @@ class SendHeadersTest(BitcoinTestFramework):
         # Create extra blocks for later
         for b in range(20):
             blocks.append(create_block(
-                tip, create_coinbase(height), block_time))
-            blocks[-1].nHeight = height
+                tip, create_coinbase(height), height, block_time))
             prepare_block(blocks[-1])
             tip = blocks[-1].sha256
             block_time += 1
@@ -575,8 +570,7 @@ class SendHeadersTest(BitcoinTestFramework):
             # Create two more blocks.
             for j in range(2):
                 blocks.append(create_block(
-                    tip, create_coinbase(height), block_time))
-                blocks[-1].nHeight = height
+                    tip, create_coinbase(height), height, block_time))
                 prepare_block(blocks[-1])
                 tip = blocks[-1].sha256
                 block_time += 1
@@ -599,8 +593,7 @@ class SendHeadersTest(BitcoinTestFramework):
         MAX_UNCONNECTING_HEADERS = 10
         for j in range(MAX_UNCONNECTING_HEADERS + 1):
             blocks.append(create_block(
-                tip, create_coinbase(height), block_time))
-            blocks[-1].nHeight = height
+                tip, create_coinbase(height), height, block_time))
             prepare_block(blocks[-1])
             tip = blocks[-1].sha256
             block_time += 1
diff --git a/test/functional/p2p_tx_download.py b/test/functional/p2p_tx_download.py
index 3203efbc0..d77257248 100755
--- a/test/functional/p2p_tx_download.py
+++ b/test/functional/p2p_tx_download.py
@@ -98,7 +98,7 @@ class TxDownloadTest(BitcoinTestFramework):
             inputs=[{
                 # coinbase
                 "txid": self.nodes[0].getblock(self.nodes[0].getblockhash(1))['tx'][0],
-                "vout": 0
+                "vout": 1
             }],
             outputs={ADDRESS_BCHREG_UNSPENDABLE: SUBSIDY - Decimal('0.025')},
         )
@@ -107,11 +107,11 @@ class TxDownloadTest(BitcoinTestFramework):
             privkeys=[self.nodes[0].get_deterministic_priv_key().key],
         )['hex']
         ctx = FromHex(CTransaction(), tx)
-        txid = int(ctx.rehash(), 16)
+        ctx.rehash()
 
         self.log.info(
             "Announce the transaction to all nodes from all {} incoming peers, but never send it".format(NUM_INBOUND))
-        msg = msg_inv([CInv(t=MSG_TX, h=txid)])
+        msg = msg_inv([CInv(t=MSG_TX, h=ctx.txid)])
         for p in self.peers:
             p.send_and_ping(msg)
 
diff --git a/test/functional/p2p_unrequested_blocks.py b/test/functional/p2p_unrequested_blocks.py
index 655d71043..a4aae0d7a 100755
--- a/test/functional/p2p_unrequested_blocks.py
+++ b/test/functional/p2p_unrequested_blocks.py
@@ -113,8 +113,7 @@ class AcceptBlockTest(BitcoinTestFramework):
         block_time = int(time.time()) + 1
         for i in range(2):
             blocks_h2.append(create_block(
-                tips[i], create_coinbase(2), block_time))
-            blocks_h2[i].nHeight = 2
+                tips[i], create_coinbase(2), 2, block_time))
             prepare_block(blocks_h2[i])
             block_time += 1
         test_node.send_and_ping(msg_block(blocks_h2[0]))
@@ -127,8 +126,7 @@ class AcceptBlockTest(BitcoinTestFramework):
 
         # 3. Send another block that builds on genesis.
         block_h1f = create_block(
-            int("0x" + self.nodes[0].getblockhash(0), 0), create_coinbase(1), block_time)
-        block_h1f.nHeight = 1
+            int("0x" + self.nodes[0].getblockhash(0), 0), create_coinbase(1), 1, block_time)
         prepare_block(block_h1f)
         block_time += 1
         test_node.send_and_ping(msg_block(block_h1f))
@@ -144,8 +142,7 @@ class AcceptBlockTest(BitcoinTestFramework):
 
         # 4. Send another two block that build on the fork.
         block_h2f = create_block(
-            block_h1f.sha256, create_coinbase(2), block_time)
-        block_h2f.nHeight = 2
+            block_h1f.sha256, create_coinbase(2), 2, block_time)
         prepare_block(block_h2f)
         block_time += 1
         test_node.send_and_ping(msg_block(block_h2f))
@@ -165,8 +162,7 @@ class AcceptBlockTest(BitcoinTestFramework):
 
         # 4b. Now send another block that builds on the forking chain.
         block_h3 = create_block(
-            block_h2f.sha256, create_coinbase(3), block_h2f.nTime + 1)
-        block_h3.nHeight = 3
+            block_h2f.sha256, create_coinbase(3), 3, block_h2f.nTime + 1)
         prepare_block(block_h3)
         test_node.send_and_ping(msg_block(block_h3))
 
@@ -192,8 +188,7 @@ class AcceptBlockTest(BitcoinTestFramework):
         for i in range(288):
             height = i + 4
             next_block = create_block(
-                tip.sha256, create_coinbase(i + 4), tip.nTime + 1)
-            next_block.nHeight = height
+                tip.sha256, create_coinbase(height), height, tip.nTime + 1)
             prepare_block(next_block)
             all_blocks.append(next_block)
             tip = next_block
@@ -271,26 +266,22 @@ class AcceptBlockTest(BitcoinTestFramework):
         # 8. Create a chain which is invalid at a height longer than the
         # current chain, but which has more blocks on top of that
         block_289f = create_block(
-            all_blocks[284].sha256, create_coinbase(289), all_blocks[284].nTime + 1)
-        block_289f.nHeight = 289
+            all_blocks[284].sha256, create_coinbase(289), 289, all_blocks[284].nTime + 1)
         prepare_block(block_289f)
 
         block_290f = create_block(
-            block_289f.sha256, create_coinbase(290), block_289f.nTime + 1)
-        block_290f.nHeight = 290
+            block_289f.sha256, create_coinbase(290), 290, block_289f.nTime + 1)
         prepare_block(block_290f)
 
         block_291 = create_block(
-            block_290f.sha256, create_coinbase(291), block_290f.nTime + 1)
-        block_291.nHeight = 291
+            block_290f.sha256, create_coinbase(291), 291, block_290f.nTime + 1)
         # block_291 spends a coinbase below maturity!
         block_291.vtx.append(create_tx_with_script(
             block_290f.vtx[0], 0, script_sig=b"42", amount=1))
         prepare_block(block_291)
 
         block_292 = create_block(
-            block_291.sha256, create_coinbase(292), block_291.nTime + 1)
-        block_292.nHeight = 292
+            block_291.sha256, create_coinbase(292), 292, block_291.nTime + 1)
         prepare_block(block_292)
 
         # Now send all the headers on the chain and enough blocks to trigger
@@ -341,8 +332,7 @@ class AcceptBlockTest(BitcoinTestFramework):
         # Now send a new header on the invalid chain, indicating we're forked
         # off, and expect to get disconnected
         block_293 = create_block(
-            block_292.sha256, create_coinbase(293), block_292.nTime + 1)
-        block_293.nHeight = 293
+            block_292.sha256, create_coinbase(293), 293, block_292.nTime + 1)
         prepare_block(block_293)
         headers_message = msg_headers()
         headers_message.headers.append(CBlockHeader(block_293))
diff --git a/test/functional/rpc_blockchain.py b/test/functional/rpc_blockchain.py
index 24915ced7..79dc818b1 100755
--- a/test/functional/rpc_blockchain.py
+++ b/test/functional/rpc_blockchain.py
@@ -332,8 +332,7 @@ class BlockchainTest(BitcoinTestFramework):
         b20 = node.getblock(b20hash)
 
         def solve_and_send_block(prevhash, height, time):
-            b = create_block(prevhash, create_coinbase(height), time)
-            b.nHeight = height
+            b = create_block(prevhash, create_coinbase(height), height, time)
             prepare_block(b)
             node.p2p.send_and_ping(msg_block(b))
             return b
diff --git a/test/functional/test_framework/blocktools.py b/test/functional/test_framework/blocktools.py
index b0b11eefe..e14ff5c18 100644
--- a/test/functional/test_framework/blocktools.py
+++ b/test/functional/test_framework/blocktools.py
@@ -28,6 +28,7 @@ from .messages import (
     CTxOut,
     FromHex,
     ToHex,
+    hash256,
 )
 from .txtools import pad_tx
 from .util import assert_equal, satoshi_round
@@ -37,7 +38,7 @@ TIME_GENESIS_BLOCK = 1600000000
 SUBSIDY = Decimal('260')
 
 
-def create_block(hashprev, coinbase, ntime=None, *, version=1):
+def create_block(hashprev, coinbase, height, ntime=None, *, version=1):
     """Create a block (with regtest difficulty)."""
     block = CBlock()
     block.nHeaderVersion = version
@@ -46,6 +47,7 @@ def create_block(hashprev, coinbase, ntime=None, *, version=1):
         block.nTime = int(time.time() + 600)
     else:
         block.nTime = ntime
+    block.nHeight = height
     block.hashPrevBlock = hashprev
     # difficulty retargeting is disabled in REGTEST chainparams
     block.nBits = 0x207fffff
diff --git a/test/functional/test_framework/messages.py b/test/functional/test_framework/messages.py
index af0f17f9c..4e912cd36 100755
--- a/test/functional/test_framework/messages.py
+++ b/test/functional/test_framework/messages.py
@@ -523,6 +523,9 @@ class CTransaction:
         txid_bytes = hash256(r)
         self.txid_hex = txid_bytes[::-1].hex()
         self.txid = uint256_from_str(txid_bytes)
+    
+    def is_coinbase(self):
+        return self.vin[0].prevout.hash == 0
 
     def input_merkle_root(self):
         hashes = []
diff --git a/test/functional/wallet_resendwallettransactions.py b/test/functional/wallet_resendwallettransactions.py
index f47bf34fa..280608709 100755
--- a/test/functional/wallet_resendwallettransactions.py
+++ b/test/functional/wallet_resendwallettransactions.py
@@ -52,8 +52,7 @@ class ResendWalletTransactionsTest(BitcoinTestFramework):
         node.setmocktime(block_time)
         height = node.getblockcount() + 1
         block = create_block(int(node.getbestblockhash(), 16), create_coinbase(
-            height), block_time)
-        block.nHeight = height
+            height), height, block_time)
         prepare_block(block)
         node.submitblock(ToHex(block))
 
-- 
2.39.5 (Apple Git-154)

